File: android/app/src/main/java/com/runformcoach/runformcoachai/sensor/RingBuffer.kt
package com.runformcoach.runformcoachai.sensor

import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

/**
 * 线程安全的泛型环形缓冲区（固定容量）。
 *
 * 容量设为 300 时，在 50 Hz 采样率下正好覆盖 6 秒窗口。
 *
 * @param T       元素类型
 * @param capacity 最大容量，构造后不可变
 *
 * 用法：
 * ```
 * val buffer = RingBuffer<SensorFrame>(capacity = 300)
 * buffer.add(frame)
 * val snapshot: List<SensorFrame> = buffer.getAll()
 * ```
 */
class RingBuffer<T>(val capacity: Int) {

    init {
        require(capacity > 0) { "容量必须大于 0，实际: $capacity" }
    }

    @Suppress("UNCHECKED_CAST")
    private val elements = arrayOfNulls<Any?>(capacity) as Array<T?>

    /** 下次写入位置，范围 [0, capacity) */
    private var writeIndex = 0

    /** 当前已存储的元素数量，范围 [0, capacity] */
    private var count = 0

    private val lock = ReentrantReadWriteLock()

    // ── 写入 ─────────────────────────────────────────────────────────────────

    /**
     * 向缓冲区追加一个元素。缓冲区满时会覆盖最旧的元素。
     */
    fun add(element: T) {
        lock.write {
            elements[writeIndex] = element
            writeIndex = (writeIndex + 1) % capacity
            if (count < capacity) {
                count++
            }
        }
    }

    // ── 读取 ─────────────────────────────────────────────────────────────────

    /**
     * 返回缓冲区中所有元素的快照，按插入顺序排列（最旧 → 最新）。
     *
     * 复杂度 O(count)，返回的是 [ArrayList] 副本，后续写入不会影响它。
     */
    fun getAll(): List<T> {
        lock.read {
            if (count == 0) return emptyList()

            val result = ArrayList<T>(count)
            // 确定读取起点：未满时从 0 开始，满时从 writeIndex（最旧元素）开始
            val start = if (count < capacity) 0 else writeIndex
            for (i in 0 until count) {
                val idx = (start + i) % capacity
                elements[idx]?.let { result.add(it) }
            }
            return result
        }
    }

    /**
     * 返回最新的 [n] 个元素（插入顺序，最旧 → 最新）。
     *
     * 若 [n] 超过当前存储量，返回所有元素。
     * 若 [n] <= 0，返回空列表。
     */
    fun getLast(n: Int): List<T> {
        require(n >= 0) { "n 不能为负: $n" }
        lock.read {
            if (count == 0 || n == 0) return emptyList()
            val take = n.coerceAtMost(count)

            val result = ArrayList<T>(take)
            val startIdx = (writeIndex - take + capacity) % capacity
            for (i in 0 until take) {
                val idx = (startIdx + i) % capacity
                elements[idx]?.let { result.add(it) }
            }
            return result
        }
    }

    // ── 辅助 ─────────────────────────────────────────────────────────────────

    /** 清空缓冲区。 */
    fun clear() {
        lock.write {
            for (i in elements.indices) {
                elements[i] = null
            }
            writeIndex = 0
            count = 0
        }
    }

    /** 当前存储的元素数量。 */
    fun size(): Int = lock.read { count }

    /** 缓冲区是否已满（已覆盖过一次）。 */
    fun isFull(): Boolean = lock.read { count == capacity }

    /** 缓冲区是否为空。 */
    fun isEmpty(): Boolean = lock.read { count == 0 }

    /** 容量。 */
    fun capacity(): Int = capacity
}

File: ios/RunFormCoachAI/CoreMotion/RingBuffer.swift
import Foundation

// MARK: - RingBuffer

/// A thread-safe, fixed-capacity circular buffer.
///
/// When full, appending a new element silently overwrites the oldest element.
/// Designed for high-throughput sensor pipelines (e.g. 100 Hz CoreMotion data).
///
/// Usage:
/// ```swift
/// var buffer = RingBuffer<SensorFrame>(capacity: 600)
/// buffer.append(frame)
/// let latest6s = buffer.all()  // up to 600 frames
/// let last100 = buffer.last(100)
/// ```
public struct RingBuffer<Element>: @unchecked Sendable {

    // MARK: - Private storage

    private final class Storage: @unchecked Sendable {
        var buffer: [Element?]
        var head: Int = 0   // next write position
        var count: Int = 0
        let capacity: Int
        let lock = os_unfair_lock_t.allocate(capacity: 1)

        init(capacity: Int, repeatedValue: Element.Type) {
            self.capacity = capacity
            self.buffer = Array<Element?>(repeating: nil, count: capacity)
            self.lock.initialize(to: os_unfair_lock())
        }

        deinit {
            lock.deinitialize(count: 1)
            lock.deallocate()
        }

        func synchronized<T>(_ block: () throws -> T) rethrows -> T {
            os_unfair_lock_lock(lock)
            defer { os_unfair_lock_unlock(lock) }
            return try block()
        }
    }

    private let storage: Storage

    // MARK: - Public properties

    /// Maximum number of elements the buffer can hold.
    public var capacity: Int { storage.capacity }

    /// Current number of elements in the buffer.
    public var count: Int {
        storage.synchronized { storage.count }
    }

    /// Whether the buffer is full (count == capacity).
    public var isFull: Bool {
        storage.synchronized { storage.count == storage.capacity }
    }

    /// Whether the buffer is empty.
    public var isEmpty: Bool {
        storage.synchronized { storage.count == 0 }
    }

    // MARK: - Init

    /// Create a ring buffer with the specified capacity.
    /// - Parameter capacity: Maximum number of elements (default 600 ≈ 6s at 100 Hz).
    public init(capacity: Int = 600) {
        precondition(capacity > 0, "RingBuffer capacity must be > 0")
        self.storage = Storage(capacity: capacity, repeatedValue: Element.self)
    }

    // MARK: - Write

    /// Append a single element. Overwrites the oldest element when the buffer is full.
    public mutating func append(_ element: Element) {
        storage.synchronized {
            storage.buffer[storage.head] = element
            storage.head = (storage.head + 1) % storage.capacity
            if storage.count < storage.capacity {
                storage.count += 1
            }
        }
    }

    /// Append a sequence of elements. Each exceeds-capacity write overwrites the oldest.
    public mutating func append(contentsOf elements: some Sequence<Element>) {
        for element in elements {
            append(element)
        }
    }

    // MARK: - Read

    /// Return all elements in FIFO order (oldest → newest).
    /// - Complexity: O(capacity).
    public func all() -> [Element] {
        storage.synchronized {
            var result: [Element] = []
            result.reserveCapacity(storage.count)
            let start = storage.count < storage.capacity
                ? 0
                : storage.head
            for i in 0..<storage.count {
                let idx = (start + i) % storage.capacity
                if let element = storage.buffer[idx] {
                    result.append(element)
                }
            }
            return result
        }
    }

    /// Return the most recent `n` elements (newest last).
    /// If `n` exceeds the current count, returns all elements.
    /// - Complexity: O(n).
    public func last(_ n: Int) -> [Element] {
        storage.synchronized {
            let limit = Swift.min(n, storage.count)
            guard limit > 0 else { return [] }
            var result: [Element] = []
            result.reserveCapacity(limit)
            // Walk backwards from (head - 1) then reverse
            for offset in 0..<limit {
                let idx = (storage.head - 1 - offset + storage.capacity) % storage.capacity
                if let element = storage.buffer[idx] {
                    result.append(element)
                }
            }
            return result.reversed()
        }
    }

    /// Return the element at the given logical index (0 = oldest, count-1 = newest).
    /// Returns `nil` if the index is out of bounds.
    public subscript(index: Int) -> Element? {
        storage.synchronized {
            guard index >= 0, index < storage.count else { return nil }
            let start = storage.count < storage.capacity
                ? 0
                : storage.head
            let physical = (start + index) % storage.capacity
            return storage.buffer[physical]
        }
    }

    /// Return the oldest element, or `nil` if empty.
    public var oldest: Element? {
        self[0]
    }

    /// Return the newest element, or `nil` if empty.
    public var newest: Element? {
        storage.synchronized {
            guard storage.count > 0 else { return nil }
            let idx = (storage.head - 1 + storage.capacity) % storage.capacity
            return storage.buffer[idx]
        }
    }

    // MARK: - Mutate

    /// Remove all elements.
    public mutating func reset() {
        storage.synchronized {
            storage.head = 0
            storage.count = 0
            // Clear references to allow deallocation
            for i in 0..<storage.capacity {
                storage.buffer[i] = nil
            }
        }
    }
}

// MARK: - Sequence conformance

extension RingBuffer: Sequence {
    public typealias Iterator = RingBufferIterator<Element>

    public func makeIterator() -> RingBufferIterator<Element> {
        RingBufferIterator(buffer: self)
    }
}

public struct RingBufferIterator<Element>: IteratorProtocol {
    private let snapshot: [Element]
    private var index = 0

    fileprivate init(buffer: RingBuffer<Element>) {
        self.snapshot = buffer.all()
    }

    public mutating func next() -> Element? {
        guard index < snapshot.count else { return nil }
        defer { index += 1 }
        return snapshot[index]
    }
}
